### DISCUSSION TITLE
photometric redshift

### DISCUSSION THREAD

**Discussion by tranthithai** [mboquien/cigale#64] [General]
I want to estimate photometric redshift using Cigale. As I understand, I should provide a negative redshift value in the input file and set a grid in the [redshiftting] mode of pcigale.ini. For example, I set a grid like: 1.5,2.0,2.5,3.0,3.5,4.0,4.5 aiming to estimate redshift in a range of 1.5< z< 4.5. However, I got an error when running : ValueError: non-broadcastable output operand with shape (1,) doesn't match the broadcast shape (18,). 

The sed order in pcigale.init : sed_modules = sfh2exp, bc03, nebular, dustatt_modified_starburst, casey2012, restframe_parameters, redshifting

does someone know the problem

**Comment by tranthithai** [comment]
# File containing the input data. The columns are 'id' (name of the
# object), 'redshift' (if 0 the distance is assumed to be 10 pc),
# 'distance' (Mpc, optional, if present it will be used in lieu of the
# distance computed from the redshift), the filter names for the fluxes,
# and the filter names with the '_err' suffix for the uncertainties. The
# fluxes and the uncertainties must be in mJy for broadband data and in
# W/m² for emission lines. Fluxes can be positive or negative. Upper
# limits are indicated with a negative value for the uncertainty. In
# case some fluxes are missing for some entries, they can be replaced
# with NaN. This file is optional to generate the configuration file, in
# particular for the savefluxes module.
data_file = MACS0723.txt

# Optional file containing the list of physical parameters. Each column
# must be in the form module_name.parameter_name, with each line being a
# different model. The columns must be in the order the modules will be
# called. The redshift column must be the last one. Finally, if this
# parameter is not empty, cigale will not interpret the configuration
# parameters given in pcigale.ini. They will be given only for
# information. Note that this module should only be used in conjonction
# with the savefluxes module. Using it with the pdf_analysis module will
# yield incorrect results.
parameters_file = 

# Available modules to compute the models. The order must be kept.
# SFH:
# * sfh2exp (double exponential)
# * sfhdelayed (delayed SFH with optional exponential burst)
# * sfhdelayedbq (delayed SFH with optional constant burst/quench)
# * sfhfromfile (arbitrary SFH read from an input file)
# * sfhperiodic (periodic SFH, exponential, rectangle or delayed)
# SSP:
# * bc03 (Bruzual and Charlot 2003)
# * m2005 (Maraston 2005; note that it cannot be combined with the nebular module)
# Nebular emission:
# * nebular (continuum and line nebular emission)
# Dust attenuation:
# * dustatt_modified_CF00 (modified Charlot & Fall 2000 attenuation law)
# * dustatt_modified_starburst (modified Calzetti 2000 attenuaton law)
# Dust emission:
# * casey2012 (Casey 2012 dust emission models)
# * dale2014 (Dale et al. 2014 dust emission templates)
# * dl2007 (Draine & Li 2007 dust emission models)
# * dl2014 (Draine et al. 2014 update of the previous models)
# * themis (Themis dust emission models from Jones et al. 2017)
# AGN:
# * skirtor2016 (AGN models from Stalevski et al. 2012, 2016)
# * fritz2006 (AGN models from Fritz et al. 2006)
# X-ray:
# * xray (from AGN and galaxies; skirtor2016/fritz2006 is needed for AGN)
# Radio:
# * radio (galaxy synchrotron emission and AGN; skirtor2016/fritz2006 is needed for AGN)
# Restframe parameters:
# * restframe_parameters (UV slope (β), IRX, D4000, EW, etc.)
# Redshift+IGM:
# * redshifting (mandatory, also includes the IGM from Meiksin 2006)
sed_modules = sfh2exp, bc03, nebular, dustatt_modified_starburst, casey2012, restframe_parameters, redshifting

# Method used for statistical analysis. Available methods: pdf_analysis,
# savefluxes.
analysis_method = pdf_analysis

# Number of CPU cores available. This computer has 56 cores.
cores = 40

# Bands to consider. To consider uncertainties too, the name of the band
# must be indicated with the _err suffix. For instance: FUV, FUV_err.
bands = jwst.nircam.F090W, jwst.nircam.F090W_err, jwst.nircam.F150W, jwst.nircam.F150W_err, jwst.nircam.F200W, jwst.nircam.F200W_err, jwst.nircam.F277W, jwst.nircam.F277W_err, jwst.nircam.F356W, jwst.nircam.F356W_err, jwst.nircam.F444W, jwst.nircam.F444W_err

# Properties to be considered. All properties are to be given in the
# rest frame rather than the observed frame. This is the case for
# instance the equivalent widths and for luminosity densities.
properties = 

# Relative error added in quadrature to the uncertainties of the fluxes
# and the extensive properties.
additionalerror = 0.1


# Configuration of the SED creation modules.
[sed_modules_params]
  
  [[sfh2exp]]
    # e-folding time of the main stellar population model in Myr.
    tau_main = 500.
    # e-folding time of the late starburst population model in Myr.
    tau_burst =  250.
    # Mass fraction of the late burst population.
    f_burst =  0.02, 0.03, 0.1, 0.15
    # Age of the main stellar population in the galaxy in Myr. The precision
    # is 1 Myr.
    age =  1, 3, 10, 30, 100, 300
    # Age of the late burst in Myr. The precision is 1 Myr.
    burst_age =  18, 20, 22, 24, 26


    # Value of SFR at t = 0 in M_sun/yr.
    sfr_0 = 1.0
    # Normalise the SFH to produce one solar mass.
    normalise = True
  
  [[bc03]]
    # Initial mass function: 0 (Salpeter) or 1 (Chabrier).
    imf = 0
    # Metalicity. Possible values are: 0.0001, 0.0004, 0.004, 0.008, 0.02,
    # 0.05.
    metallicity =  0.004
    # Age [Myr] of the separation between the young and the old star
    # populations. The default value in 10^7 years (10 Myr). Set to 0 not to
    # differentiate ages (only an old population).
    separation_age = 10
  
  [[nebular]]
    # Ionisation parameter. Possible values are: -4.0, -3.9, -3.8, -3.7,
    # -3.6, -3.5, -3.4, -3.3, -3.2, -3.1, -3.0, -2.9, -2.8, -2.7, -2.6,
    # -2.5, -2.4, -2.3, -2.2, -2.1, -2.0, -1.9, -1.8, -1.7, -1.6, -1.5,
    # -1.4, -1.3, -1.2, -1.1, -1.0.
    logU = -3.0, -2.5, -2.2, -2.1, -2.0, -1.9, -1.8, -1.5, -1.3, -1.1
    # Gas metallicity. Possible values are: 0.000, 0.0004, 0.001, 0.002,
    # 0.0025, 0.003, 0.004, 0.005, 0.006, 0.007, 0.008, 0.009, 0.011, 0.012,
    # 0.014, 0.016, 0.019, 0.020, 0.022, 0.025, 0.03, 0.033, 0.037, 0.041,
    # 0.046, 0.051.
    zgas = 0.014, 0.02, 0.025, 0.041
    # Electron density. Possible values are: 10, 100, 1000.
    ne = 100
    # Fraction of Lyman continuum photons escaping the galaxy. Possible
    # values between 0 and 1.
    f_esc = 0.0
    # Fraction of Lyman continuum photons absorbed by dust. Possible values
    # between 0 and 1.
    f_dust = 0.0
    # Line width in km/s.
    lines_width = 300.0
    # Include nebular emission.
    emission = True
  
  [[dustatt_modified_starburst]]
    # E(B-V)l, the colour excess of the nebular lines light for both the
    # young and old population.
    E_BV_lines = 0.0, 0.1, 0.15, 0.2, 0.25, 0.3
    # Reduction factor to apply on E_BV_lines to compute E(B-V)s the stellar
    # continuum attenuation. Both young and old population are attenuated
    # with E(B-V)s.
    E_BV_factor = 0.3, 0.35, 0.4
    # Central wavelength of the UV bump in nm.
    uv_bump_wavelength = 217.5
    # Width (FWHM) of the UV bump in nm.
    uv_bump_width = 35.0
    # Amplitude of the UV bump. For the Milky Way: 3.
    uv_bump_amplitude = 0.0
    # Slope delta of the power law modifying the attenuation curve.
    powerlaw_slope = 0.0
    # Extinction law to use for attenuating the emissio  n lines flux.
    # Possible values are: 1, 2, 3. 1: MW, 2: LMC, 3: SMC. MW is modelled
    # using CCM89, SMC and LMC using Pei92.
    Ext_law_emission_lines = 1
    # Ratio of total to selective extinction, A_V / E(B-V), for the
    # extinction curve applied to emission lines.Standard value is 3.1 for
    # MW using CCM89, but can be changed.For SMC and LMC using Pei92 the
    # value is automatically set to 2.93 and 3.16 respectively, no matter
    # the value you write.
    Rv = 3.1
    # Filters for which the attenuation will be computed and added to the
    # SED information dictionary. You can give several filter names
    # separated by a & (don't use commas).
    filters = B_B90 & V_B90 & FUV
  
  [[casey2012]]
    # Temperature of the dust in K.
    temperature = 35.0
    # Emissivity index of the dust.
    beta = 1.6
    # Mid-infrared powerlaw slope.
    alpha = 2.0
  
  [[restframe_parameters]]
    # Observed and intrinsic UV slopes β and β₀ measured in the same way as
    # in Calzetti et al. (1994).
    beta_calz94 = True
    # D4000 break using the Balogh et al. (1999) definition.
    D4000 = False
    # IRX computed from the GALEX FUV filter and the dust luminosity.
    IRX = False
    # Central wavelength of the emission lines for which to compute the
    # equivalent width. The half-bandwidth must be indicated after the '/'
    # sign. For instance 656.3/1.0 means oth the nebular line and the
    # continuum are integrated over 655.3-657.3 nm.
    EW_lines = 500.7/1.0 & 656.3/1.0
    # Filters for which the rest-frame luminosity will be computed. You can
    # give several filter names separated by a & (don't use commas).
    luminosity_filters = FUV & V_B90
    # Rest-frame colours to be computed. You can give several colours
    # separated by a & (don't use commas).
    colours_filters = FUV-NUV & NUV-r_prime
  
  [[redshifting]]
    # Redshift of the objects. Leave empty to use the redshifts from the
    # input file.
    redshift = 1.25,1.30,1.35,1.40,1.45,1.50,1.60,1.65,1.70,1.75,1.80,1.85


# Configuration of the statistical analysis method.
[analysis_params]
  # List of the physical properties to estimate. Leave empty to analyse
  # all the physical properties (not recommended when there are many
  # models).
  variables = sfh.sfr, sfh.sfr10Myrs, sfh.sfr100Myrs
  # List of bands for which to estimate the fluxes. Note that this is
  # independent from the fluxes actually fitted to estimate the physical
  # properties.
  bands = jwst.nircam.F090W, jwst.nircam.F150W, jwst.nircam.F200W, jwst.nircam.F277W, jwst.nircam.F356W, jwst.nircam.F444W
  # If true, save the best SED for each observation to a file.
  save_best_sed = False
  # Save the raw chi2. It occupies ~15 MB/million models/variable. Allowed
  # values are 'all', 'none', 'properties', and 'fluxes'.
  save_chi2 = none
  # Take into account upper limits. If 'full', the exact computation is
  # done. If 'noscaling', the scaling of the models will not be adjusted
  # but the χ² will include the upper limits adequately. Waiving the
  # adjustment makes the fitting much faster compared to the 'full' option
  # while generally not affecting the results in any substantial manner.
  # This is the recommended option as it achieves a good balance between
  # speed and reliability. Finally, 'none' simply discards bands with
  # upper limits.
  lim_flag = noscaling
  # If true, for each object we create a mock object and analyse them.
  mock_flag = False
  # When redshifts are not given explicitly in the redshifting module,
  # number of decimals to round the observed redshifts to compute the grid
  # of models. To disable rounding give a negative value. Do not round if
  # you use narrow-band filters.
  redshift_decimals = 2
  # Number of blocks to compute the models and analyse the observations.
  # If there is enough memory, we strongly recommend this to be set to 1.
  blocks = 1

**Reply by mboquien** [reply]
Hi. Can you attach the files (pasting the `pcigale.ini` makes it a little difficult to read) along with a one-object test-case so I try to reproduce the issue? Thanks!

**Reply by tranthithai** [reply]
[pcigale.txt](https://github.com/mboquien/cigale/files/12443552/pcigale.txt)

**Reply by mboquien** [reply]
I do not see anything particularly suspicious in the `pcigale.ini` file. Can you provide the full backtrace so I know where the error is happening? Thanks.

**Reply by tranthithai** [reply]
Dear Boquien. Please see the attach file to see the warning
[warning.txt](https://github.com/mboquien/cigale/files/12487808/warning.txt)

**Reply by mboquien** [reply]
Oh I see. The issue is that you have a `burst_age` that is older than the `age`:
```
    age =  1, 3, 10, 30, 100, 300
    # Age of the late burst in Myr. The precision is 1 Myr.
    burst_age =  18, 20, 22, 24, 26
```
You must ensure that the smallest `age` is larger than the largest `burst_age`. Give the range of redshift you are using, you may want to adopt much larger ages in any case.

**Reply by tranthithai** [reply]
omg. You are right. The problem is solved. Thank you very much.
